### PR TITLE
feat: olive updates, more scaffolding, and basic 'make test'

### PR DESCRIPTION
**Pull request by kdmccormick**
## Overview

This PR updates the cookiecutter to use the latest hooks from Tutor v15 (Olive). It also adds scaffolding for both custom CLI commands and custom jobs ("do-commands") , which I expect many plugin authors will be interested in. Furthemore, it un-comments several sections of the cookiecutter that were previously commented-out in order to bring cookiecutter users as close to a working plugin as possible. Finally, it adds some basic static checks via two Makefiles: one in the cookiecutter repo root, and another in the root of the generated plugin.

## Specific changes

* Confirm that cookiecutter generates a plugin that is compatible
  with the Tutor v15 (and Open edX Olive).
  Closes https://github.com/overhangio/cookiecutter-tutor-plugin/issues/12.

* Replace COMMANDS_INIT (which is deprecated in v15/Olive) with
  equivalent usage of CLI_DO_INIT_TASKS (which is new in v15/Olive).
  Closes https://github.com/overhangio/cookiecutter-tutor-plugin/issues/15.

* Add boilerplate & explanation for CLI_DO_COMMANDS (new in v15/Olive),
  include a dummy example `tutor local do say-hi` job.

* Add boilerplate and explanation for CLI_COMMANDS, including
  a dummy example `tutor myplugin print-repo` command.

* Un-comment as much of the image management block as possible
  in order to help avoid instances of syntactically invalid commented-out
  code such as: https://github.com/overhangio/cookiecutter-tutor-plugin/pull/16

* Add Makefile to both cookiecutter root as well as generated plugin,
  providing make rules `test`, `test-format`, `test-lint`, and
  `test-types`. When running CI from the cookiecutter, we use sed to
  un-comment the commented code blocks in order to increase check
  coverage. These checks are not yet run in CI.
  See https://github.com/overhangio/cookiecutter-tutor-plugin/issues/7.

## Example of generated plugin.py
```python
from __future__ import annotations

import os
import os.path
from glob import glob

import click
import pkg_resources
from tutor import hooks

from .__about__ import __version__

########################################
# CONFIGURATION
########################################

hooks.Filters.CONFIG_DEFAULTS.add_items(
    [
        # Add your new settings that have default values here.
        # Each new setting is a pair: (setting_name, default_value).
        # Prefix your setting names with 'MYPLUGIN_'.
        ("MYPLUGIN_VERSION", __version__),
    ]
)

hooks.Filters.CONFIG_UNIQUE.add_items(
    [
        # Add settings that don't have a reasonable default for all users here.
        # For instance: passwords, secret keys, etc.
        # Each new setting is a pair: (setting_name, unique_generated_value).
        # Prefix your setting names with 'MYPLUGIN_'.
        # For example:
        ### ("MYPLUGIN_SECRET_KEY", "{{ 24|random_string }}"),
    ]
)

hooks.Filters.CONFIG_OVERRIDES.add_items(
    [
        # Danger zone!
        # Add values to override settings from Tutor core or other plugins here.
        # Each override is a pair: (setting_name, new_value). For example:
        ### ("PLATFORM_NAME", "My platform"),
    ]
)


########################################
# INITIALIZATION TASKS
########################################

# To add a custom initialization task, create a bash script template under:
# tutormyplugin/templates/myplugin/jobs/init/
# and then add it to the MY_INIT_TASKS list. Each task is in the format:
# ("<service>", ("<path>", "<to>", "<script>", "<template>"))
MY_INIT_TASKS: list[tuple[str, tuple[str, ...]]] = [
    # For example, to add LMS initialization steps, you could add the script template at:
    # tutormyplugin/templates/myplugin/jobs/init/lms.sh
    # And then add the line:
    ### ("lms", ("myplugin", "jobs", "init", "lms.sh")),
]


# For each task added to MY_INIT_TASKS, we load the task template
# and add it to the CLI_DO_INIT_TASKS filter, which tells Tutor to
# run it as part of the `init` job.
for service, template_path in MY_INIT_TASKS:
    full_path: str = pkg_resources.resource_filename(
        "tutormyplugin", os.path.join("templates", *template_path)
    )
    with open(full_path, encoding="utf-8") as init_task_file:
        init_task: str = init_task_file.read()
    hooks.Filters.CLI_DO_INIT_TASKS.add_item((service, init_task))


########################################
# DOCKER IMAGE MANAGEMENT
########################################


# Images to be built by `tutor images build`.
# Each item is a quadruple in the form:
#     ("<tutor_image_name>", ("path", "to", "build", "dir"), "<docker_image_tag>", "<build_args>")
hooks.Filters.IMAGES_BUILD.add_items(
    [
        # To build `myimage` with `tutor images build myimage`,
        # you would add a Dockerfile to templates/myplugin/build/myimage,
        # and then write:
        ### (
        ###     "myimage",
        ###     ("plugins", "myplugin", "build", "myimage"),
        ###     "docker.io/myimage:{{ MYPLUGIN_VERSION }}",
        ###     (),
        ### ),
    ]
)


# Images to be pulled as part of `tutor images pull`.
# Each item is a pair in the form:
#     ("<tutor_image_name>", "<docker_image_tag>")
hooks.Filters.IMAGES_PULL.add_items(
    [
        # To pull `myimage` with `tutor images pull myimage`, you would write:
        ### (
        ###     "myimage",
        ###     "docker.io/myimage:{{ MYPLUGIN_VERSION }}",
        ### ),
    ]
)


# Images to be pushed as part of `tutor images push`.
# Each item is a pair in the form:
#     ("<tutor_image_name>", "<docker_image_tag>")
hooks.Filters.IMAGES_PUSH.add_items(
    [
        # To push `myimage` with `tutor images push myimage`, you would write:
        ### (
        ###     "myimage",
        ###     "docker.io/myimage:{{ MYPLUGIN_VERSION }}",
        ### ),
    ]
)


########################################
# TEMPLATE RENDERING
# (It is safe & recommended to leave
#  this section as-is :)
########################################

hooks.Filters.ENV_TEMPLATE_ROOTS.add_items(
    # Root paths for template files, relative to the project root.
    [
        pkg_resources.resource_filename("tutormyplugin", "templates"),
    ]
)

hooks.Filters.ENV_TEMPLATE_TARGETS.add_items(
    # For each pair (source_path, destination_path):
    # templates at ``source_path`` (relative to your ENV_TEMPLATE_ROOTS) will be
    # rendered to ``source_path/destination_path`` (relative to your Tutor environment).
    # For example, ``tutormyplugin/templates/myplugin/build``
    # will be rendered to ``$(tutor config printroot)/env/plugins/myplugin/build``.
    [
        ("myplugin/build", "plugins"),
        ("myplugin/apps", "plugins"),
    ],
)


########################################
# PATCH LOADING
# (It is safe & recommended to leave
#  this section as-is :)
########################################

# For each file in tutormyplugin/patches,
# apply a patch based on the file's name and contents.
for path in glob(
    os.path.join(
        pkg_resources.resource_filename("tutormyplugin", "patches"),
        "*",
    )
):
    with open(path, encoding="utf-8") as patch_file:
        hooks.Filters.ENV_PATCHES.add_item((os.path.basename(path), patch_file.read()))


########################################
# CUSTOM JOBS (a.k.a. "do-commands")
########################################

# A job is a set of tasks, each of which run inside a certain container.
# Jobs are invoked using the `do` command, for example: `tutor local do importdemocourse`.
# A few jobs are built in to Tutor, such as `init` and `createuser`.
# You can also add your own custom jobs:

# To add a custom job, define a Click command that returns a list of tasks,
# where each task is a pair in the form ("<service>", "<shell_command>").
# For example:
### @click.command()
### @click.option("-n", "--name", default="plugin developer")
### def say_hi(name: str) -> list[tuple[str, str]]:
###     """
###     An example job that just prints 'hello' from within both LMS and CMS.
###     """
###     return [
###         ("lms", f"echo 'Hello from LMS, {name}!'"),
###         ("cms", f"echo 'Hello from CMS, {name}!'"),
###     ]


# Then, add the command function to CLI_DO_COMMANDS:
## hooks.Filters.CLI_DO_COMMANDS.add_item(say_hi)

# Now, you can run your job like this:
#   $ tutor local do say-hi --name="John Doe"


#######################################
# CUSTOM CLI COMMANDS
#######################################

# Your plugin can also add custom commands directly to the Tutor CLI.
# These commands are run directly on the user's host computer
# (unlike jobs, which are run in containers).

# To define a command group for your plugin, you would define a Click
# group and then add it to CLI_COMMANDS:


### @click.group()
### def myplugin() -> None:
###     pass
### 
### 
### hooks.Filters.CLI_COMMANDS.add_item(myplugin)


# Then, you would add subcommands directly to the Click group, for example:


### @myplugin.command()
### def example_command() -> None:
###     """
###     This is helptext for an example command.
###     """
###     print("You've run an example command.")


# This would allow you to run:
#   $ tutor myplugin example-command

```

### CODE REVIEW

**Comment by kdmccormick** [comment]
@overhangio/tutor-maintainers Here are some improvements to the plugin cookiecutter that are ready for review! This is not critical in order for the cookiecutter to work with Olive, but it does include an update to remove the `COMMANDS_INIT` filter (which is deprecated in Olive) so I'd like to merge this within a week or two if possible.

File: Makefile
Comment (kdmccormick) [inline Makefile:6]:
```suggestion
# We will add them to CI as part of https://github.com/overhangio/cookiecutter-tutor-plugin/issues/7.
```

File: {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py
Comment (regisb) [inline {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py:71]: There is no need to wrap the scripts in `bash -c "..."`.  Scripts are already wrapped in `sh -e -c "..."`: https://github.com/overhangio/tutor/blob/nightly/tutor/commands/compose.py#L388
Also, bash may not be available in the container. (only sh)

File: {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py
Comment (regisb) [inline {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py:191]: Great example!

File: {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py
Comment (regisb) [inline {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py:195]: Do we want to leave this uncommented by default? I suspect that most people will not take the time to delete this command and will end up with a "say_hi" job in every one of their plugins. (As a sidenote, these identical jobs _will_ generate conflicts...)

Same comment below for the custom CLI command. (which will not conflict but I think that most plugin developers don't need extra custom commands)

File: {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py
Comment (regisb) [inline {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py:57]: We should probably not have these init tasks in the plugin by default. First it might be confusing for the developer. Second, running every one of these tasks will take a handful of seconds, even if the task doesn't do anything. (that's because running the lms/cms containers triggers a restart of the permission services)

**Review by regisb — CHANGES_REQUESTED** [review]
I think that we should try to keep the plugin that is generated by default pretty lean. I mean, it's great that the plugin contains extensive instructions on how to use some of the important hooks, but we should avoid the creation of no-op init tasks and commands. Maybe just comment out the calls to `hooks.Filters.<filter>.add_item`?

File: {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py
Comment (kdmccormick) [reply {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py:71]: 👍🏻

File: {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py
Comment (kdmccormick) [reply {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py:195]: > I suspect that most people will not take the time to delete this command and will end up with a "say_hi" job in every one of their plugins. (As a sidenote, these identical jobs will generate conflicts...)

Yeah, I hear you.

My concern with the current setup is that there's no syntax checking being done on any of the lines that are commented out. I try to be diligent by manually checking the syntax of the commented-out lines whenever I change them, but [one error](https://github.com/overhangio/cookiecutter-tutor-plugin/pull/16) already slipped past me. Further, it's not just syntax correctness I want to check for: I want the cookie-cut plugin to pass linting, black and mypy [so that we can turn those on by default for new plugins](https://github.com/overhangio/cookiecutter-tutor-plugin/issues/7).

There's another way I could do it, though. I could comment out every line containing code with two `#`s:
```python
# To add a custom job, define a Click command that returns a list of tasks,
# where each task is a pair in the form ("<service>", "<shell_command>").
# For example:
##@click.command()
##@click.option("-n", "--name", default="plugin developer")
##def say_hi(name: str) -> list[tuple[str, str]]:
##    """
##    An example job that just prints 'hello' from within both LMS and CMS.
##    """
##   return [
##        ("lms", f"echo 'Hello from LMS, {name}!'"),
##        ("cms", f"echo 'Hello from CMS, {name}!'"),
##    ]
```

Then, this repository could check the generated plugin.py by substituting `"##"` for `""` and then running pylint/black/mypy on the resulting file.

Comment (regisb) [reply {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py:195]: Let me wander in philosophy territory for a minute... I'm perfectly OK with bugs slipping into the source code, as long as we do our best to detect and fix them quickly enough. Two pairs of eyes (yours and mine) are enough to find  issues reported in the forum and fix them. In my experience, trying to preemptively protect ourselves from all sorts of bugs leads to over-complicated processes and engineering.

It's not a black/white thing of course, it's a balance that we need to strike. If you feel like you need the automated syntax checking for every change in the codebase, then sure go for it. As far as I'm concerned, the syntax checking that happens in my IDE is enough for me.

You're the maintainer, so in the end you know what's best. You should just know that no one is going to blame you if there are mistakes in your code. People who don't do anything don't break anything either :)

Comment (kdmccormick) [reply {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py]: > In my experience, trying to preemptively protect ourselves from all sorts of bugs leads to over-complicated processes and engineering.

Yes, this is something I have found myself susceptible to, and I appreciate you pushing towards simplicity. In retrospect, complicating the *output* of the cookiecutter for the sake of automatic verifiability would have been putting the cart before the horse.

I do feel, though, that keeping the example code in the cookiecutter in a working state is pretty important, since it sits at sensitive point in the Tutor & Open edX adoption funnel: when someone uses it, they might be embarking on the journey to make their first plugin, which I see as a big step towards becoming a Tutor power user and an active community member. If the example code doesn't work, they might move throw their hands up in frustration and move onto something else! (I mean, hopefully, they would ask for help and then our lovely community would help them :) but people are busy, and they might instead just move on to the next thing because they need a quick solution.)

I've pushed another commit, which returns all the code to a commented-out state, but uses a special prefix (`### `) which this repo's Makefile removes when it runs tests. This way, the complexity is relegated to the CI suite of this repo, and shouldn't leak out to the cookiecutter users. Let me know what you think--if it seems wrong, I'm happy to revert it.

Comment (regisb) [reply {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py]: I appreciate your desire to have correct code, and I think your solution is terrific.

File: Makefile
Comment (kdmccormick) [reply Makefile:17]: Great trick.

**Pull request merged**
